### PR TITLE
[REVIEW] BUG Move subcomms init outside of individual algorithm functions

### LINKED ISSUE #1065: [FEA] sub-communicator initialization for 2D partitioning support
**Is your feature request related to a problem? Please describe.**
In 2D partitioning, we align P workers (i.e. GPUs) to P_row * P_column and also partition the graph adjacency matrix in 2D.

Common communication patterns are collectives among the workers in the same row or column. To support this, we need to add sub-communicators for the workers in the same row and the same column.

https://github.com/rapidsai/raft/pull/18 and https://github.com/rapidsai/raft/pull/44 added sub-communicator support to RAFT comms.

cuGraph currently initializes only the global communicator. We need to initialize sub-communicaors as well.

https://github.com/rapidsai/raft/blob/f93dad05574b84d32ebbbd25681d2f9bcd7c0a14/cpp/include/raft/comms/comms.hpp#L95
```comm_split``` (similar to MPI_Comm_split https://www.mpich.org/static/docs/latest/www3/MPI_Comm_split.html) splits the global communicator to sub-communicators.

https://github.com/rapidsai/raft/blob/f93dad05574b84d32ebbbd25681d2f9bcd7c0a14/cpp/include/raft/handle.hpp#L163
https://github.com/rapidsai/raft/blob/f93dad05574b84d32ebbbd25681d2f9bcd7c0a14/cpp/include/raft/handle.hpp#L167
```set_subcomm``` and ```get_subcomm``` can be used add sub-communicators to the handle and retrieve the added sub-communicator when necessary.

- [x] Decide P_row & P_column for the given P. We may set P_row and P_column as close as possible to the square root of P by default with a user option to override the default value.
- [x] Call comm_split to create sub-communicators (two sub-communicators per worker, one for collectives among the workers in the same row, and the other for collectives among the workers in the same column)
- [x] Decide the string keys for the row sub-communicator and column sub-communicator.
- [x] Using the string keys, register the sub-communicators to the RAFT handle (call set_subcomm)

### PR DESCRIPTION
**Pull request by rlratzel**
This PR moves the subcomms init outside of the individual algorithm functions and up to the function a user calls to initialize comms. This allows the same subcomm configuration to be used across algorithms and also allows users to customize the call to optimize based on GPU configuration.

Without this, subcomms will be repeatedly created/initialized for each algo call, resulting in problems.

closes #1065

(thanks to @Iroy30 for commits https://github.com/rapidsai/cugraph/pull/1196/commits/71d4a0b652060d8bf365599028af3653d87b6838 and https://github.com/rapidsai/cugraph/pull/1196/commits/79f131c567ab9ccd8ecc601d60cb58c85b3aec9c)

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by rlratzel** [comment]
closes #1065

File: python/cugraph/comms/comms.py [hunk 1]
```diff
@@ -14,15 +14,65 @@
...
+    __subcomm = (prows, pcols, partition_type)
+
+
+def _subcomm_init(sID, partition_row_size):
+    handle = get_handle(sID)
+    c_init_subcomms(handle, partition_row_size)
 
 
 # Intialize Comms. If explicit Comms not provided as arg,
 # default Comms are initialized as per client information.
```
Comment (seunghwak) [inline python/cugraph/comms/comms.py:70]: As this is a public interface, don't we need to update the documentation?

**Review by seunghwak — APPROVED** [review]
Besides the comment about documentation, this looks good to me. Thank you for the quick fix!

**Review by ChuckHastings — APPROVED** [review]
Much cleaner!

File: python/cugraph/comms/comms.py (hunk 1 above)
Comment (seunghwak) [reply python/cugraph/comms/comms.py:70]: And we eventually need to remove the C++ partition manager class. This class is basically a misnomer... C++ is not managing partitioning. The python layer is managing partitioning and the C++ layer is just performing a passive task based on instructions from the python layer.

Comment (rlratzel) [reply python/cugraph/comms/comms.py:70]: Yes, thank you. I'll work on that now.

File: CHANGELOG.md [hunk 2]
```diff
@@ -39,8 +39,9 @@
 - PR #1166 Fix misspelling of function calls in asserts causing debug build to fail
 - PR #1180 BLD Adopt RAFT model for cuhornet dependency
 - PR #1181 Fix notebook error handling in CI
-- PR #1186 BLD Installing raft headers under cugraph 
+- PR #1186 BLD Installing raft headers under cugraph
 - PR #1192 Fix benchmark notes and documentation issues in graph.py
+- PR #1196 Move subcomms init outside of individual algorithm functions
```
Comment (afender) [inline CHANGELOG.md:44]: Consider using new enhanced tags for automated reports:

```
ENH       Enhancement, new functionality
BUG       Bug Fix
DOC       Additions/updates to documentation
TST         Additions/updates to tests
BLD        Updates to the build process/scripts
PRF        Performance improvement
CLN        Code cleanup
```

File: cpp/include/utilities/cython.hpp [hunk 3]
```diff
@@ -201,5 +199,8 @@ std::pair<size_t, weight_t> call_louvain(raft::handle_t const& handle,
                                          size_t max_level,
                                          weight_t resolution);
 
+// Helper for setting up subcommunicators
+void init_subcomms(raft::handle_t& handle, size_t row_comm_size);
```
Comment (afender) [inline cpp/include/utilities/cython.hpp:203]: Public API header right? 
Users will wonder what `row_comm_size` is, how it should be set. Any specific default behavior should be documented as well.

File: python/cugraph/comms/comms.py [hunk 4]
```diff
@@ -114,6 +174,10 @@ def get_worker_id(sID):
     return sessionstate['wid']
 
 
-def get_n_workers(sID):
-    sessionstate = worker_state(sID)
-    return sessionstate['nworkers']
+def get_n_workers(sID=None):
```
Comment (afender) [inline python/cugraph/comms/comms.py:177]: I could sware I saw several ways in the python layer that already get the number of workers/GPUs:
- get_n_workers() (from read_utils)
- len(get_visible_devices())
- len(numba.cuda.gpus)

 We should document why this one is different/better or avoid multiplying code doing the same thing.

File: cpp/include/utilities/cython.hpp (hunk 3 above)
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp:203]: Thanks, working on that now for the next commit.

File: python/cugraph/comms/comms.py (hunk 4 above)
Comment (rlratzel) [reply python/cugraph/comms/comms.py:177]: Agreed, I'll look into this for the next commit.

**Review by afender — APPROVED** [review]

**Comment by rlratzel** [comment]
rerun tests

**Pull request merged**